### DISCUSSION TITLE
Fibaro FGT001 multilevel sensor reports

### DISCUSSION THREAD

**Discussion by kvaster** [zwave-js/zwave-js#3427] [Q&A]
Two months ago I've made PR for fgt001 for v8. This PR ignores endpoint 1 for fgt001, cause endpoint 1 is a subset of root endpoint. And endpoint 2 is used for multilevel sensor and second battery (external sensor). I thought it was working correct with such patch, but it seems no.

The problem is with reports for notifications and values. After node reinterview I can see in groups following settings:
```
0 - Lifeline - zwave stick - 0
2 - LIfeline - zwave stick - None
```

With such settings reports for setpoint changes and notifications reports are sent from endpoint 1. As a result I can see additional setpoint and notification entries in zwavejs2mqtt. Reports for temerature settings are sent automatically from time to time (on change) as expected from endpoint 2.

First I've tried to remove all associations and create them in the following way:
```
0 - Lifeline - zwave stick - None
2 - LIfeline - zwave stick - None
```

With such settings setpoint changes and notifications reports are sent from endpoint 0. But there is NO any automatic reports for temperature changes (generally all changes on endpoint 2).

Also I've tried folloing settings:
```
0 - Lifeline - zwave stick - 0
2 - LIfeline - zwave stick - 0
```

Such settings will result in reports from endpoint 1 and also absense of automatic reports on endpoint 2.

Settings 0/None or None/0 manually is not possible - it always leads to 0/0 after refreshing associations.

Hiding endpoint 0 insteadof 1 is also bad idea, cause thermostat setpoint is not working properly in such case - probably it is cause endpoint 1 is just a subset of root endpoint.

For now I'm confused - don't know how to make all reports working properly... The best idea I can think of is to implement endpoints merging...

ps: using latest node-zwave-js version as of Sep 28, 2021 with zwavejs2mqtt now.

**Comment by AlCalzone** [comment]
Have you tried asking the Fibaro support how the device's associations are supposed to be configured?

I'm starting to suspect that this device actually needs to use endpoints 1&2 instead of 0&2. 
To try this theory, can you edit your config file to contain this:
```
"preserveEndpoints": [1, 2]
```
then re-interview and post a log of that?

This should leave you with the following associations:
```
1 - Lifeline - zwave stick - None
2 - Lifeline - zwave stick - None
```
and hopefully working reports.

**Comment by kvaster** [comment]
In docs for this thermostat I can only see mention of Endpoint 1 and Endpoint 2. Tried [1,2] and "*" settings and with them I can see following associations by default:

```
0 - Lifeline - zwave-stick - 0
1 - Lifeline - zwave-stick - None
2 - Lifeline - zwave-stick - 0
```

Also I can see three batteries with such settings: endpoint 0 duplicates endpoint 1.

[zwavejs_2021-09-29.log](https://github.com/zwave-js/node-zwave-js/files/7252914/zwavejs_2021-09-29.log)

Testing right now if endpoint 2 reports are working...

**Reply by AlCalzone** [reply]
I'm a bit confused why endpoints 1 and 2 have different lifeline targets (one with an endpoint association [0], one with a node association [none]). I would have expected both to use node associations, but in your log it says that they are associated with an endpoint association via the root endpoint (which should only be a fallback solution if nothing else worked).
Do you have `associations` configured in the config file?

**Reply by kvaster** [reply]
No. There are no any `associations` configured in the config file.

**Reply by AlCalzone** [reply]
I might need to add some debug logging to get to the bottom of this. Are you able to run zwavejs2mqtt against a different branch of zwave-js like it is described here?
https://zwave-js.github.io/zwavejs2mqtt/#/development/custom-docker

**Reply by kvaster** [reply]
Yes. I'm running zwavejs2mqtt against latest git branch right now. And can make fast changes.

**Reply by kvaster** [reply]
Also changing anything in associations manually leads to no reports on endpoint 2.

**Reply by AlCalzone** [reply]
Ok, I'll try to add something later and will ping you on the PR. Got some other stuff to do first, so don't wait for me.

**Reply by kvaster** [reply]
Thanks a lot.
Also - battery duplication is OK ? Should I just add compat flag to remove battery class on endpoint 0 in this case ?

**Reply by AlCalzone** [reply]
We'll deal with the battery duplication later. Here's the test PR: https://github.com/zwave-js/node-zwave-js/pull/3431
The bot will also release a test build in case that's easier for you to build against.

Please re-interview your device with that build and show me the log output.

**Comment by kvaster** [comment]
Logs with lib from PR:

[zwavejs_2021-09-30.log](https://github.com/zwave-js/node-zwave-js/files/7257766/zwavejs_2021-09-30.log)

You can see two nodes in this log: 18 and 24. I found some differences between them. I've added node 24 (by accident) in security mode and it shows S0_Legacy right now and by default node 24 have all targets set to '0'.

ps: also it seems there are some differences with discovering batteries in security mode.

**Comment by AlCalzone** [comment]
Ok let's try once more, PR is building. It seems that the Multi Channel Association groups of the endpoints are shared with the root, so they cannot be configured separately. The normal associations might not be, but the interview didn't query them.

I've re-ordered two commands, so we'll hopefully see this now.

**Comment by kvaster** [comment]
Now for node 18 i can see:
```
0 | Lifeline | zwave (home) | None
0 | Lifeline | zwave (home) | 0
1 | Lifeline | zwave (home) | None
2 | Lifeline | zwave (home) | None
```

And for node 24:
```
0 | Lifeline | zwave (home) | 0
1 | Lifeline | zwave (home) | 0
2 | Lifeline | zwave (home) | 0
```

Logs:
[zwavejs_2021-09-30.log](https://github.com/zwave-js/node-zwave-js/files/7258187/zwavejs_2021-09-30.log)

**Comment by AlCalzone** [comment]
Okay, this one for Node 18 is probably not there anymore. I can see it get removed, but the cached association doesn't get updated properly:
```
0 | Lifeline | zwave (home) | 0
```
The normal associations of the endpoints are also shared with the root.

Node 24 is different, but that is due to the interview order. By the point the associations for the root endpoint are set up, the secure capabilities of the other endpoints are not know yet, so the driver thinks they don't support Association CC. As a result, the root is configured with an endpoint association instead of a classic association.

---

Since you now have the only two possible combinations of associations: Is one of the nodes now reporting on its own correctly? For now I don't care whether endpoint 0 or 1 gets the reports, as long as it is reliable.

**Reply by kvaster** [reply]
I'm testing right now... Meanwhile I've tried to remove all associations and to reinterview nodes. Node 24 - no changes. And node 18:

```
0 | Lifeline | zwave (home) | None
0 | Lifeline | zwave (home) | 0
1 | Lifeline | zwave (home) | None
2 | Lifeline | zwave (home) | 0
```

Endpoint 2 target changes to '0' (and is always same after removing all associations and interview).
And there are always two associations on endpoint 0... Refreshing them does not help.

**Reply by AlCalzone** [reply]
Can you share a log of you removing the associations and then refreshing them?

**Comment by kvaster** [comment]
Log:
[zwavejs_2021-09-30.log](https://github.com/zwave-js/node-zwave-js/files/7259858/zwavejs_2021-09-30.log)

* Removed all associations
* Refreshed them
* Node reinterview

Done for both 18 and 24 nodes.

Also node 24 is receiving endpoint 2 reports while node 18 doesn't in such setup.

**Reply by AlCalzone** [reply]
So, so node 18 now should have a node association on the root endpoint, which is shared with the endpoints
```
0 | Lifeline | zwave (home) | None
1 | Lifeline | zwave (home) | None
2 | Lifeline | zwave (home) | None
```

Node 24 should have a multi channel association that is also shared
```
0 | Lifeline | zwave (home) | 0
1 | Lifeline | zwave (home) | 0
2 | Lifeline | zwave (home) | 0
```

I can't see unsolicited reports in your logs, only two manual refreshes at the end. But your comment sounds like Node 24 is behaving correctly, while 18 isn't?
What about Endpoint 1? Does it report as itself or via the root endpoint?

**Reply by kvaster** [reply]
Updates: It looks I was wrong. I can see unsolicited reports in logs for both node 18 and node 24. Also reports are sent on endpoint 1 for both nodes.

Also if I do 'remove all associations', 'refresh them' (no associations shown in this moment in zwavejs2mqtt panel), 'reinterview node' then for node 18 I can see:


```
0 | Lifeline | zwave (home) | None
0 | Lifeline | zwave (home) | 0
1 | Lifeline | zwave (home) | None
2 | Lifeline | zwave (home) | 0
```

And for 24:

```
0 | Lifeline | zwave (home) | 0
1 | Lifeline | zwave (home) | 0
2 | Lifeline | zwave (home) | 0
```

**Reply by kvaster** [reply]
Forget to mention that reports for second battery and for temperature sensor are always coming from endpoint 2.

**Reply by AlCalzone** [reply]
I still don't understand why endpoint 1 is different than endpoint 2 for node 18. 

Does everything work now if you ignore endpoint 0? We will need an additional compat flag to hide endpoint 0 in setups like this one, so you'll have to live with the extraneous values for now.

**Reply by kvaster** [reply]
Is this test PR going to be merged ?

For now only battery is duplicated and I can use patched version of device config:

```
        "compat": {
                "preserveEndpoints": [1,2],
                "mapRootReportsToEndpoint": 1,
                "commandClasses": {
                        "remove": {
                                "0x80": {
                                        "endpoints": [0]
                                }
                        }
                }
        }
```

The only thing I don't understand is battery on node 24 (with S0_Legacy security). I can see battery statuses and values after node reinterview, but battery class is completely removed (on all endpoints) after zwavejs2mqtt restart. This behavious occurs only if 'remove' compact flag exists and only with S0_Legacy security.

Thanks a lot for helping.

**Reply by AlCalzone** [reply]
> Is this test PR going to be merged ?

If it works for you

> The only thing I don't understand is battery on node 24 (with S0_Legacy security).

Can you share a log of a re-interview?

**Reply by kvaster** [reply]
Log:
[zwavejs_2021-10-01.log](https://github.com/zwave-js/node-zwave-js/files/7267583/zwavejs_2021-10-01.log)

Reinterview is done with config from my previous message. See no batteries in zwavejs2mqtt after restart.

**Reply by AlCalzone** [reply]
The queries are successful and values are created. @robertsLando any idea why z2m is not showing those values after restart?
How can we debug this?

**Reply by robertsLando** [reply]
We need to see zwavejs2mqtt logs and also see a node json dump

**Reply by kvaster** [reply]
The problem is inside Node.getDefinedValueIDs - this function doesn't return battery values on node ready. I'm currently investigating problem further.

**Reply by kvaster** [reply]
I've moved deeper into code...
In values db:

```
{"k":"{\"nodeId\":24,\"commandClass\":96,\"endpoint\":1,\"property\":\"commandClasses\"}","v":[94,152,159,34]}
```
and
```
{"k":"{\"nodeId\":18,\"commandClass\":96,\"endpoint\":1,\"property\":\"commandClasses\"}","v":[94,142,64,67,83,152,159,89,108,113,117,128,129,133,34]}
```

**Reply by AlCalzone** [reply]
Can you share both the `<homeid>.json` and the value DB? If necessary, you can send it to the email in my profile.
I'm not sure right now why that is even stored in the value DB, the supported CCs of endpoints are tracked in the `<homeid>.json` file.

**Reply by kvaster** [reply]
Nothing secret in these files:

[f71d83d3.json](https://github.com/zwave-js/node-zwave-js/files/7276088/f71d83d3.json.txt)
[f71d83d3.values.jsonl](https://github.com/zwave-js/node-zwave-js/files/7276090/f71d83d3.values.jsonl.txt)

There are no battery CCs at all in <homeid>.json file...

**Reply by AlCalzone** [reply]
Just FYI, I won't be able to look at these until next Sunday

**Reply by kvaster** [reply]
Thanks! I'm going to investigate problem further on my own for now.

**Reply by AlCalzone** [reply]
Did you find anything out?
If you can step through with a debugger, you can put breakpoints in `src/lib/node/Node.ts` in the methods `applyCommandClassesCompatFlag` and `overwriteConfig` and check what happens there when you re-interview the device.
If not, I can add some debug logging there.

**Reply by kvaster** [reply]
Hello. No. Had too little time unfortunatelly.
Thanks for the tips on `applyCommandClassesCompatFlag` - I will try to debug this in a day or two. I had to readd this thermostat to the system, cause it began to show me errors like not responding to GetV1 request (don't rememer exacly right now) and interview started to always fail... I have one more unused thermostat - will add it to system in S2 secure mode and will try to debug code.

Another two strange things on my side:

For thermostats in non-secure mode if I will remove associations and reinterview them I have following associations:

```
0 | Lifeline | zwave (home) | None
0 | Lifeline | zwave (home) | 0
1 | Lifeline | zwave (home) | None
2 | Lifeline | zwave (home) | 0
```

But if I will reinterview them without removing associations then I will have another associations:

```
0 | Lifeline | zwave (home) | None
0 | Lifeline | zwave (home) | 0
1 | Lifeline | zwave (home) | None
2 | Lifeline | zwave (home) | None
```

So, it seems that final associations after reinterview really depends on clearing or not associations before starting reinterview.

And secon strange thing is that after some time I can see 'corrupted' additional values in multisensors for some devices. For example, I can see voltage and watts in some of my CO2 sensors, or CO2 value in some of my thermostats. Is it possible that data may be corrupted with bad signal with some of devices ?

**Reply by AlCalzone** [reply]
I'm relatively sure that some of the associations with target endpoint 0 don't actually exist. I've found one situation during the interview where the associations weren't refreshed: https://github.com/zwave-js/node-zwave-js/pull/3490

> after some time I can see 'corrupted' additional values in multisensors for some devices

These are corrupted when receiving due to interference or something. I need to add filtering of sensor types that aren't reported as supported. FWIW, this is not an issue with S0/S2 because these corrupted commands can't be decrypted.

**Reply by AlCalzone** [reply]
> it began to show me errors like not responding to GetV1 request

I've seen that somewhere else today, but that is because the Version CC wasn't exposed insecurely.

**Comment by kvaster** [comment]
About PR: it looks like it working almost same like before. The only difference is associations I can see in zwavejs2mqtt panel.

**Reply by AlCalzone** [reply]
So, good to merge?

**Reply by kvaster** [reply]
For my setup this PR breaks nothing. I think yes - good to merge.
